### PR TITLE
FS-4895: When creating a fund or a round check, whether given round shortnames are available

### PR DESCRIPTION
**Pull request by nuwan-samarasinghe**
**Ticket: https://mhclgdigital.atlassian.net/browse/FS-4895**

### Change description
In this change we are focussing on fund and round short name reuse issue so following scenarios are acceptable 

1. If there is a fund and that given short name is not in our records that short name is acceptable 
2. If there is an update for the given fund short name is acceptable because it belongs to it self 
3. IF there is a fund and when round creation if the short name is not available then acceptable 
4. IF there is a different fund and we are planing to use same short name it is acceptable 
5. If we do update the round then it will be again acceptable since short name belongs to it self

Following are not acceptable 

1. If there is a fund for a given short name then we are shoving an error for validation purpose 
2. If there is a fund and inside that fund if we have same short name round we give an error 

- [x] Unit tests and other appropriate tests added or updated
- [x] README and other documentation has been updated / added (if needed)
- [x] Commit messages are meaningful and follow good commit message guidelines (e.g. "FS-XXXX: Add margin to nav items preventing overlapping of logo")

### How to test

**Testing Fund create**

1. Use FAB 
2. Try to create a fund that does not exists then it should show the success 
3. Try to create a existing short named fund it gives error
4. Try to rename a short name of a fund to a already existing one then it should give an error other wise should be success

**Testing Round Create**

1. Use FAB
2. Try to rename a short name of a round to a already existing one then it should give an error other wise should be success

### Screenshots of UI changes (if applicable)

image

image

### CODE REVIEW

File: app/blueprints/fund_builder/routes.py [hunk 1]
```diff
@@ -254,42 +254,50 @@ def fund(fund_id=None):
...
-        add_fund(new_fund)
-        flash(f"Created fund {form.name_en.data}")
-        return redirect(url_for(BUILD_FUND_BP_INDEX))
+            add_fund(new_fund)
+            flash(f"Created fund {form.name_en.data}")
+            return redirect(url_for(BUILD_FUND_BP_INDEX))
 
-    error = error_formatter(form.errors)
+        error = {**error,
+                 'short_name': [f'Given fund short name already exists.']}
```
Comment (samuelhwilliams) [inline app/blueprints/fund_builder/routes.py:299]: this doesn't seem familiar to me as GOV.UK error message content style - do we want to run this through content design? Or is that all happening separately?

**Review by wjrm500 — CHANGES_REQUESTED** [review]
Hey @nuwan-samarasinghe, I think a better implementation would be to create validation functions and attach them directly to the FlaskForms e.g.,

In `apps/funding-service-design-fund-application-builder/app/blueprints/fund_builder/forms/fund.py`:

```
...
def unique_fund_short_name(form, field):
    fund = get_fund_by_short_name(field.data)
    if fund and str(fund.fund_id) != form.fund_id.data:
        raise ValidationError("Given fund short name already exists.")
...
class FundForm(FlaskForm):
    ...
    short_name = StringField(
        "Short name",
        validators=[
            DataRequired(),
            Length(max=10),
            no_spaces_between_letters,
            unique_fund_short_name,
        ],
    )
    ...
```

This way we don't need to modify the route function or the error formatter at all, and we achieve the same result.

We can then do the same for Round. Let me know what you reckon!

File: app/blueprints/fund_builder/routes.py (hunk 1 above)
Comment (nuwan-samarasinghe) [reply app/blueprints/fund_builder/routes.py:298]: We have a ticket to verify the error message with content designer @chaneldiep

**Review by NarenderRajuB — APPROVED** [review]
LGTM

**Review by NarenderRajuB — DISMISSED** [review]
LGTM

File: tests/test_routes.py [hunk 2]
```diff
@@ -115,6 +147,66 @@ def test_update_fund(flask_test_client, seed_dynamic_data):
...
+        "feedback_link": "http://example.com/feedback",
+        "project_name_field_id": 1,
+        "guidance_url": "http://example.com/guidance",
+    }
+
+    response = submit_form(flask_test_client, "/round", new_round_data)
+    assert response.status_code == 200
+    new_round_data = {**new_round_data, 'short_name': 'NR1234'}
+    response = submit_form(flask_test_client, "/round", new_round_data)
+    assert response.status_code == 200
```
Comment (wjrm500) [inline tests/test_routes.py:203]: Could we have assertions to check the error message doesn't appear in the HTML for the first two submissions? As we expect a 200 response either way. We could have:

```
error_html = '<a href="#short_name">Short name: Given short name already exists in the fund funding to improve testing.</a>'

# Test works fine with first round
response = submit_form(flask_test_client, "/round", new_round_data)
assert response.status_code == 200
assert error_html not in response.data.decode('utf-8'), "Error HTML found in response"

# Test works fine with second round but with different short name
new_round_data = {**new_round_data, 'short_name': 'NR1234'}
response = submit_form(flask_test_client, "/round", new_round_data)
assert response.status_code == 200
assert error_html not in response.data.decode('utf-8'), "Error HTML found in response"

# Test doesn't work with third round with same short name as firsrt
new_round_data = {**new_round_data, 'short_name': 'NR123'}
response = submit_form(flask_test_client, "/round", new_round_data)
assert response.status_code == 200
assert error_html in html = response.data.decode('utf-8'), "Error HTML not found in response"
```

**Review by wjrm500 — CHANGES_REQUESTED** [review]
Just a couple of small changes and we're good to go!

**Review by wjrm500 — APPROVED** [review]

**Pull request merged**